spec.bs

```diff
@@ -289,46 +318,179 @@ To <dfn for="PrivateAggregation">obtain an aggregatable report</dfn> given an [=
-1. [=set/iterate|For each=] |contribution| of |truncatedContributions|:
-    <!-- 1. [=Assert=]: |contribution| is a {{PAHistogramContribution}}. -->
-    1. [=Assert=]: |contribution|["|value|"] is non-negative.
-    1. Add |value| to |contributionSum|.
-1. Let |sufficientBudget| be the result of [=PrivateAggregation/consuming budget if permitted=] given |contributionSum|, |reportingOrigin| and |api|.
-1. If |sufficientBudget| is false, return.
-1. Let |report| be the result of [=PrivateAggregation/obtaining an aggregatable report=] given |reportingOrigin|, |api| and |truncatedContributions|.
-1. [=set/Append=] |report| to the user agent's [=aggregatable report cache=].
+Note: This section is largely copied from the <a href="https://wicg.github.io/attribution-reporting-api/">Attribution Reporting API spec</a>, adapting as necessary.
 
-Issue: Do we need to ensure the reports aren't queued after being sent?
+Issue: Add logic for resetting [=aggregatable report/queued=] on start up or network on or whatever
 
-Issue: Do we need to address user settings here at all?
+Issue: Do we have to use the [=queue a task=] algorithm here?
 
-Issue: Does the [=max stored reports=] part match the implementation?
+The user agent must periodically [=attempt to queue reports for sending=] with its [=aggregatable report cache=].
 
-To <dfn algorithm for="PrivateAggregation">consume budget if permitted</dfn> given a {{long}} |value|, an [=origin=] <var ignore=''>origin</var> and an [=context type=] |api|, perform [=implementation-defined=] steps. They return a [=boolean=], which indicates whether there is sufficient 'contribution budget' left to send the requested contribution |value|. This budget should be bound to usage over time, e.g. the contribution sum over the last 24 hours. The algorithm should assume that the contribution will be sent if and only if true is returned.
+To <dfn>attempt to queue reports for sending</dfn> given a [=list=] of [=aggregatable reports=] |reports|:
+1. [=list/iterate|For each=] |report| of |reports|, run these steps [=in parallel=]:
+    1. Run these steps, but [=abort when=] the [=user agent=] shuts down:
+        1. If |report|'s [=aggregatable report/queued=] value is true, return.
+        1. Set |report|'s [=aggregatable report/queued=] value to true.
+        1. If |report|'s [=aggregatable report/report time=] is less than the [=wall clock=]'s [=current unsafe time=], set |report|'s [=aggregatable report/report time=] to [=wall clock=]'s [=current unsafe time=] plus an [=implementation-defined=] random non-negative [=duration=].
 
-Issue: Don't we need to take into account the time "now" as well?
+            Note: On startup, it is possible the user agent will need to send many reports whose report times passed while the browser was closed. Adding random delay prevents temporal joining of reports.
+        1. Wait until the [=wall clock=]'s [=current unsafe time=] is equal to |report|'s [=aggregatable report/report time=].
```

Comment:
Is it ok to compare equality for a coarsen time and an unsafe time?

Reply:
My reading is it might be, but we may as well change to "current wall time" anyway.

Two follow up questions: should we change all of these from |context|'s [=current wall time=] to just the general [=current wall time=]? I'm also curious if the clock we're using (i.e. base::Time::Now()) actually does this coarsening. Do you know?

side note: I've changed this to "equal to or after", but lmk if you preferred it before

Reply:
@apasel422 made the change for the wall time for ARA and should know more about this. @apasel422 Do you know whether it's preferred to use context's current wall time or the general current wall time? ARA uses context's current wall time throughout. Thanks.

Reply:
No, I don't know.

Reply:
The context's current wall time may have higher precision in [cross-origin-isolated contexts](https://html.spec.whatwg.org/multipage/webappapis.html#concept-settings-object-cross-origin-isolated-capability). Does this apply to shared storage or fledge worklet? If so, maybe better to use the context's current wall time.

Reply:
Just given that we use base::Time::Now() throughout, my instinct would be to match that 'global' state in the spec (unless we also want to change the implementation)

Reply:
Yeah I'm fine with that, just keeping it consistent in the spec.

Reply:
Done. I'll try and follow up with the spec mentor about this too